pallets/messages/src/lib.rs

```diff
@@ -165,6 +167,49 @@ pub mod pallet {
 
 	#[pallet::call]
 	impl<T: Config> Pallet<T> {
+		/// Adds a message for a resource hosted on IPFS. The IPFS payload should
+		/// contain both a
+		/// [CID](https://docs.ipfs.io/concepts/content-addressing/#identifier-formats)
+		/// as well as a 32-bit payload length.
+
+		/// # Arguments
+		/// * `origin` - A signed transaction origin from the provider
+		/// * `on_behalf_of` - Optional. The msa id of delegate.
+		/// * `schema_id` - Registered schema id for current message.
+		/// * `cid` - The content address for an IPFS payload
+		/// * `payload_length` - The size of the payload
+		/// * Returns
+		/// * [DispatchResultWithPostInfo](https://paritytech.github.io/substrate/master/frame_support/dispatch/type.DispatchResultWithPostInfo.html)
+		#[pallet::weight(T::WeightInfo::add_ipfs_message(payload.cid.get().len() as u32, 1_000))]
+		pub fn add_ipfs_message(
+			origin: OriginFor<T>,
+			on_behalf_of: Option<MessageSourceId>,
+			schema_id: SchemaId,
+			payload: IPFSPayload,
+		) -> DispatchResultWithPostInfo {
+			let provider_key = ensure_signed(origin)?;
+
+			let schema = T::SchemaProvider::get_schema_by_id(schema_id);
+			ensure!(schema.is_some(), Error::<T>::InvalidSchemaId);
+			ensure!(
+				schema.unwrap().payload_location == PayloadLocation::IPFS,
+				Error::<T>::InvalidPayloadLocation
+			);
+
+			let message_source_id = Self::find_msa_id(&provider_key, on_behalf_of)?;
+			let message = Self::add_message(
+				provider_key,
+				message_source_id,
+				payload.cid.get().clone().try_into().unwrap(),
```

Comment:
I think we want to handle the error here. Perhaps doing something similar from `add_on_chain_message`

Reply:
what error? payload size or possible null pointer from the unwrap?

Reply:
For the cid exceeding the size of the bounded vector.

Reply:
Done.